File: AdMobMediation_v8/ViewController/SuprAdBanner/SuprAdBannerViewController.m
//
//  SuprAdBannerViewController.m
//  GoogleMediation
//
//  Created by JustinTsou on 2021/4/23.
//

#import "SuprAdBannerViewController.h"
#import <GoogleMobileAds/GoogleMobileAds.h>
#import <AotterTrek-iOS-SDK/AotterTrek-iOS-SDK.h>

static NSString *const TestSuprAdUnit_BottomBanner = @"ca-app-pub-8836593984677243/5995562909";

@interface SuprAdBannerViewController ()<GADNativeAdLoaderDelegate>
@property (weak, nonatomic) IBOutlet UIButton *refreshButton;
@property (atomic, strong) GADAdLoader *adLoader;
@property (nonatomic, strong) GADNativeAd *gADUnifiedSuprAd_BottomBanner;
@property (nonatomic, strong) UIView *suprAdBackgroundView;
@end

@implementation SuprAdBannerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupGADAdLoader];
}

#pragma mark : Button Action GADAdLoader

- (IBAction)tapRefreshButton:(UIButton *)sender {
    [self adLoaderLoadRequest];
}

#pragma mark : Setup GADAdLoader

- (void)setupGADAdLoader {

    self.adLoader = [[GADAdLoader alloc]initWithAdUnitID: TestSuprAdUnit_BottomBanner
                                      rootViewController: self
                                                 adTypes: @[kGADAdLoaderAdTypeNative]
                                                 options: @[]];
    
    self.adLoader.delegate = self;

    [self adLoaderLoadRequest];
}

- (void)adLoaderLoadRequest {
    [self.adLoader loadRequest:[GADRequest request]];
}

#pragma mark - GADUnifiedNativeAdLoaderDelegate

- (void)adLoader:(GADAdLoader *)adLoader didReceiveNativeAd:(GADNativeAd *)nativeAd {

    if (nativeAd != nil) {

        if ([[nativeAd.extraAssets allKeys]containsObject:@"trekAd"]) {
            NSString *adType = nativeAd.extraAssets[@"trekAd"];
            
            if ([adType isEqualToString:@"suprAd"]) {
                // Get nativeAd
                _gADUnifiedSuprAd_BottomBanner = nativeAd;
                
                
                // Setup Banner UI
                if ([[_gADUnifiedSuprAd_BottomBanner.extraAssets allKeys]containsObject:@"adSizeWidth"] &&
                    [[_gADUnifiedSuprAd_BottomBanner.extraAssets allKeys]containsObject:@"adSizeHeight"]) {
                    
                    // get ad prefered AdSize
                    NSString *width = _gADUnifiedSuprAd_BottomBanner.extraAssets[@"adSizeWidth"];
                    NSString *height = _gADUnifiedSuprAd_BottomBanner.extraAssets[@"adSizeHeight"];
                    double adSizeWidth = [width doubleValue];
                    double adSizeHeight = [height doubleValue];
                    
                    CGFloat viewWidth = UIScreen.mainScreen.bounds.size.width;
                    CGFloat viewHeight = viewWidth * adSizeHeight/adSizeWidth;
                    int adheight = (int)viewHeight;
                    CGSize preferedMediaViewSize = CGSizeMake(viewWidth, adheight);
                    
                    [self setupSuprAdBackgroundView:_gADUnifiedSuprAd_BottomBanner preferedMediaViewSize:preferedMediaViewSize];
                }
            }
        }
    }
}

- (void)adLoader:(GADAdLoader *)adLoader didFailToReceiveAdWithError:(NSError *)error {
    NSLog(@"Error Message:%@",error.description);
}


#pragma mark - Private Method

- (void)setupSuprAdBackgroundView:(GADNativeAd *)nativeAd preferedMediaViewSize:(CGSize)size {
    
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    _suprAdBackgroundView = [[UIView alloc]initWithFrame:rect];
    
    [self.view addSubview:_suprAdBackgroundView];
    [_suprAdBackgroundView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    [_suprAdBackgroundView.heightAnchor constraintEqualToConstant:size.height].active = YES;
    [_suprAdBackgroundView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor].active = YES;
    [_suprAdBackgroundView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor constant:0].active = YES;
    [_suprAdBackgroundView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor constant:0].active = YES;
    
    
    GADMediaView *gADMediaView = [[GADMediaView alloc]initWithFrame:CGRectZero];
    gADMediaView.mediaContent = nativeAd.mediaContent;
    [_suprAdBackgroundView addSubview:gADMediaView];

    [gADMediaView setTranslatesAutoresizingMaskIntoConstraints:NO];

    [gADMediaView.topAnchor constraintEqualToAnchor:_suprAdBackgroundView.topAnchor constant:0].active = YES;
    [gADMediaView.bottomAnchor constraintEqualToAnchor:_suprAdBackgroundView.bottomAnchor constant:0].active = YES;
    [gADMediaView.leftAnchor constraintEqualToAnchor:_suprAdBackgroundView.leftAnchor constant:0].active = YES;
    [gADMediaView.rightAnchor constraintEqualToAnchor:_suprAdBackgroundView.rightAnchor constant:0].active = YES;
}

@end

File: AdMobMediation_v8_Swift/ViewController/SuprAdBanner/SuprAdBannerViewController.swift
//
//  SuprAdBannerViewController.swift
//  AotterGoogleMediationAdSwift
//
//  Created by JustinTsou on 2021/4/23.
//

import UIKit
import GoogleMobileAds

class SuprAdBannerViewController: UIViewController {
    
    private let TestSuprAdUnit_BottomBanner: String = "ca-app-pub-8836593984677243/5995562909"
    
    @IBOutlet weak var refreshButton: UIButton!
    private var adLoader: GADAdLoader?
    private var gADUnifiedSuprAd_BottomBanner: GADNativeAd?
    private var suprAdBackgroundView: UIView?

    override func viewDidLoad() {
        super.viewDidLoad()

        self.setupGADAdloader()
    }

    @IBAction func tapRefreshButton(_ sender: UIButton) {
        self.adLoaderLoadRequest()
    }
}

// MARK: Setup GADAdLoader

extension SuprAdBannerViewController {
    
    private func setupGADAdloader() {
        
        adLoader = GADAdLoader.init(adUnitID: TestSuprAdUnit_BottomBanner, rootViewController: self, adTypes: [.native], options: [])

        adLoader?.delegate = self

        adLoaderLoadRequest()
    }
    
    private func adLoaderLoadRequest() {
        adLoader?.load(GADRequest())
    }
}

// MARK: GADUnifiedNativeAdLoaderDelegate

extension SuprAdBannerViewController: GADNativeAdLoaderDelegate {

    func adLoader(_ adLoader: GADAdLoader, didReceive nativeAd: GADNativeAd) {
        guard let extraAssets = nativeAd.extraAssets else { return }

        if extraAssets.keys.contains("trekAd") {
            let adType = extraAssets["trekAd"] as! String

            if adType == "suprAd" {
                gADUnifiedSuprAd_BottomBanner = nativeAd
                
                if extraAssets.keys.contains("adSizeWidth") &&
                    extraAssets.keys.contains("adSizeHeight") {
                    
                    let width = extraAssets["adSizeWidth"] as! Double
                    let height = extraAssets["adSizeHeight"] as! Double
                    
                    let adSizeWidth:CGFloat = CGFloat(width)
                    let adSizeHeight:CGFloat = CGFloat(height)
                    
                    let viewWidth = UIScreen.main.bounds.size.width
                    let viewHeight = viewWidth * adSizeHeight/adSizeWidth
                    let adheight = Int(viewHeight)
                    let adWidth = Int(viewWidth)
                    let preferedMediaViewSize = CGSize(width: adWidth, height: adheight)
                    
                    self.setupSuprAdBackgroundView(nativeAd: gADUnifiedSuprAd_BottomBanner, size: preferedMediaViewSize)
                }
            }
        }
    }

    func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: Error) {
        print("Error Message:\(error.localizedDescription)")
    }
}

extension SuprAdBannerViewController {
    private func setupSuprAdBackgroundView(nativeAd:GADNativeAd?, size:CGSize) {
        let rect = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        self.suprAdBackgroundView = UIView(frame: rect)
        
        self.view.addSubview(self.suprAdBackgroundView ?? UIView())
        self.suprAdBackgroundView?.translatesAutoresizingMaskIntoConstraints = false
        
        self.suprAdBackgroundView?.heightAnchor.constraint(equalToConstant: size.height).isActive = true
        self.suprAdBackgroundView?.bottomAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.bottomAnchor).isActive = true
        self.suprAdBackgroundView?.trailingAnchor.constraint(equalTo: self.view.trailingAnchor).isActive = true
        self.suprAdBackgroundView?.leadingAnchor.constraint(equalTo: self.view.leadingAnchor).isActive = true
        
        let gADMediaView:GADMediaView = GADMediaView(frame: .zero)
        gADMediaView.mediaContent = nativeAd?.mediaContent
        self.suprAdBackgroundView?.addSubview(gADMediaView)
        
        gADMediaView.translatesAutoresizingMaskIntoConstraints = false
        
        gADMediaView.topAnchor.constraint(equalTo: self.suprAdBackgroundView!.topAnchor).isActive = true
        gADMediaView.bottomAnchor.constraint(equalTo: self.suprAdBackgroundView!.bottomAnchor).isActive = true
        gADMediaView.trailingAnchor.constraint(equalTo: self.suprAdBackgroundView!.trailingAnchor).isActive = true
        gADMediaView.leadingAnchor.constraint(equalTo: self.suprAdBackgroundView!.leadingAnchor).isActive = true
    }
}
